src/liboslexec/batched_backendllvm.cpp

```diff
@@ -811,26 +811,29 @@ BatchedBackendLLVM::llvm_load_value(llvm::Value* ptr, const TypeSpec& type,
         // If it's an array or we're dealing with derivatives, step to the
         // right element.
         TypeDesc t = type.simpletype();
+        // TODO: may be the wrong type
+        llvm::Type* element_type = llvm_type(t.elementtype());
         if (t.arraylen || deriv) {
             int d = deriv * std::max(1, t.arraylen);
             llvm::Value* elem;
             if (arrayindex)
                 elem = ll.op_add(arrayindex, ll.constant(d));
             else
                 elem = ll.constant(d);
-            ptr = ll.GEP(ptr, elem);
+            ptr = ll.GEP(element_type, ptr, elem);
         }
 
         // If it's multi-component (triple or matrix), step to the right field
         if (!type.is_closure_based() && t.aggregate > 1) {
             OSL_DEV_ONLY(std::cout << "step to the right field " << component
                                    << std::endl);
-            ptr = ll.GEP(ptr, 0, component);
+            ptr = ll.GEP(element_type, ptr, 0, component);
         }
 
         // Now grab the value
-        llvm::Value* result;
-        result = ll.op_load(ptr);
+        // TODO: fails with assert later in ll.native_to_llvm_mask
```

Comment:
Well, here we are relying on the underlying llvm type to identify if the value we just loaded is uniform or varying.
This differs from the parameter "bool op_is_uniform" which indicates if the op the participating in is varying or not.  I.E.:
`[[varying]] int C = [[uniform]] A != [[varying]] B;`

In that case the op itself is varying, and when `A` is loaded, it will need to be widened before being passed to LLVM_Util::op_eq(...) so it matches types with the varying value loaded for `B`.

So for batched, not only do we need to know the TypeSpec of the symbol being loaded, we will also need an explicit "bool symbol_is_uniform".  In this case if that parameter was added:
```
llvm::Value*
BatchedBackendLLVM::llvm_load_value(llvm::Value* ptr, const TypeSpec& type,
                                    int deriv, llvm::Value* arrayindex,
                                    int component, TypeDesc cast,
                                    bool symbol_is_uniform, bool op_is_uniform, bool index_is_uniform,
                                    bool symbol_forced_boolean)
```
Then the code that was causing problems (asserting):
```
        if (type.is_int() && symbol_forced_boolean) {
            // We only need to convert wide native masks
            // and op_is_uniform doesn't guarantee that the symbol it self
            // in non-unform, it could just be a single bool (vs. mask).
            if (!op_is_uniform && (ll.llvm_typeof(result) != ll.type_bool())) {
                // We just loaded a native mask need to convert it
                // to a vector of bools
                result = ll.native_to_llvm_mask(result);
            }
```
could be written as:
```
        if (type.is_int() && symbol_forced_boolean) {
            // We only need to convert wide native masks
            // and op_is_uniform doesn't guarantee that the symbol it self
            // in non-unform, it could just be a single bool (vs. mask).
            if (!op_is_uniform && symbols_is_uniform) {
                // We just loaded a native mask need to convert it
                // to a vector of bools
                result = ll.native_to_llvm_mask(result);
            }
```            
You see how the previous use of the llvm::Value *result's type,  `(ll.llvm_typeof(result) != ll.type_bool())`, was just to figure out if the boolean value loaded was uniform or varying.

Of course this means having all callers pass a valid symbols_is_uniform parameter in, luckily the `Symbol::is_uniform()` has the answer.

So would suggest starting with that change and use the extensive testsuite to route out other useses.
Your might even comment out LLVM_Util::llvm_typeof to route out all its uses.

Reply:
Thanks! That really helped me understand things better.